src/main/java/org/broadinstitute/hellbender/utils/bwa/BwaMemIndex.java

```diff
@@ -0,0 +1,349 @@
+package org.broadinstitute.hellbender.utils.bwa;
+
+import htsjdk.samtools.Cigar;
+import htsjdk.samtools.CigarElement;
+import htsjdk.samtools.CigarOperator;
+import htsjdk.samtools.SAMFlag;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.NativeUtils;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.List;
+
+public final class BwaMemIndex implements AutoCloseable {
+    private long indexAddress;
+    private final List<String> refContigNames;
+    private static volatile BwaMemIndex globalInstance;
+
+    public static void createIndexImage( final String refName, final String imgName ) {
+        assertNonEmptyReadable(refName+".amb");
+        assertNonEmptyReadable(refName+".ann");
+        assertNonEmptyReadable(refName+".bwt");
+        assertNonEmptyReadable(refName+".pac");
+        assertNonEmptyReadable(refName+".sa");
+        createIndexImageFile(refName, imgName);
+    }
+
+    public static BwaMemIndex getInstance( final String indexImageFile ) {
+        BwaMemIndex result = globalInstance;
+        if ( result == null ) {
+            synchronized(BwaMemIndex.class) {
+                result = globalInstance;
+                if ( result == null ) globalInstance = result = new BwaMemIndex(indexImageFile);
+            }
+        }
+        return result;
+    }
+
+    public static void closeInstance() {
+        BwaMemIndex result = globalInstance;
+        if ( result != null ) {
+            synchronized(BwaMemIndex.class) {
+                result = globalInstance;
+                if ( result != null ) {
+                    globalInstance = null;
+                    result.close();
+                }
+            }
+        }
+    }
+
+    public static void closeAllDistributedInstances( final JavaSparkContext ctx ) {
+        int nJobs = ctx.defaultParallelism();
```

Comment:
I've moved the 3 static methods that deal with a singleton instance into a separate class, BwaMemIndexSingleton.  Now BwaMemIndex knows nothing about Spark, and can also be used locally.

I've added a check in BwaMemIndexSingleton.getInstance() that we don't already have a different index loaded.  This will at least make us aware of abandonment in some cases (when we're switching indexes), and will also prevent people from accidentally writing one pipeline that tries to use two different indexes.

Best I can do, for now, I think.